mlos_bench/mlos_bench/services/base_service.py

```diff
@@ -79,13 +81,34 @@ def __init__(self,
         self.config = config or {}
         self._validate_json_config(self.config)
         self._parent = parent
-        self._services: Dict[str, Callable] = {}
+        self._service_methods: Dict[str, Callable[[Self, Any], Any]] = {}
 
         if parent:
             self.register(parent.export())
         if methods:
             self.register(methods)
 
+        # In order to get a list of all child contexts, we need to look at only
+        # the bound methods that were not overridden by another mixin.
+        # Then we inspect the internally bound __self__ variable to discover
+        # which Service instance that method belongs too.
+        # To do this we also
+
+        # As a sanity check, make sure all methods registered are bound.
+        assert all(hasattr(svc_method, '__self__') and isinstance(svc_method.__self__, Service)
+                   for svc_method in self._service_methods.values())
```

Comment:
Yeah, I'd already adjusted those to make them methods, but it is a reasonable point.
Initially I was thinking to use that as a sanity test to make sure I wasn't missing something.
But if we want I could use it as a filter instead so that only service instances are considered for context entry.
I'm not sure there'd be a use case for non class instances for that.

Reply:
Reverted that and changed it to just a filter.